training/src/app/modules/chuck-norris/components/search/search.component.ts

```diff
@@ -0,0 +1,81 @@
+import { Component, Output, EventEmitter, OnInit } from '@angular/core';
+import { AbstractControl, FormBuilder, FormGroup, Validators } from "@angular/forms";
+import { debounceTime, Subject, takeUntil, tap } from "rxjs";
+
+@Component({
+	selector: 'app-search',
+	templateUrl: './search.component.html',
+	styleUrls: ['./search.component.scss'],
+
+})
+export class SearchComponent implements OnInit {
+
+	public searchForm: FormGroup;
+	private searchKeyIsValid: boolean;
+	private componentDestroyed$: Subject<void>                        = new Subject();
+	@Output() private generateJokesOnClickMode: EventEmitter<string>  = new EventEmitter<string>();
+	@Output() private generateJokesOnSearchMode: EventEmitter<string> = new EventEmitter<string>();
+	@Output() private isSearchValid: EventEmitter<boolean>            = new EventEmitter<boolean>();
+
+	constructor(private fb: FormBuilder) {
+	}
+
+	ngOnInit(): void {
+		this.searchForm = this.fb.group({
+			searchKey: [
+				'', [
+					Validators.required,
+					Validators.minLength(3),
+					Validators.maxLength(120)
+				]
+			],
+		});
+		this.getJokesWhenSearchKeyChange();
+	}
+
+	get searchKeyFormControl(): AbstractControl {
+		return this.searchForm.get('searchKey');
+	}
+
+	public getJokesWhenSearchKeyChange(): void {
+		this.searchKeyFormControl?.valueChanges.pipe(
+			tap((searchValue: string) => {
+				this.searchKeyIsValid = this.searchKeyFormControl?.valid
+				this.SearchKeyValidation(searchValue)
+				}
+			),
+			debounceTime(1000),
+			takeUntil(this.componentDestroyed$)
+		).subscribe((searchValue: string) => {
+			if (this.searchKeyIsValid) {
+				this.setItems(searchValue);
+			}
+		});
+	}
+
+	public SearchKeyValidation(searchValue:string): void {
+		if (this.searchKeyIsValid) {
+			this.isSearchValid.emit(false);
+			this.generateJokesOnClickMode.emit(searchValue);
```

Comment:
the user wrote something in the input, how is that related to a "clickedMode" event?